src/utils/packageUtils.ts

```diff
@@ -0,0 +1,976 @@
+  path: '',
+  package: '',
+  versionName: 'ver 0.1',
+  versionNumber: '0.1.0.NEXT',
+  default: true,
+};
+
+const logger = Logger.childFromRoot('packageUtils');
+export const BY_PREFIX = ((): IdRegistry => {
+  return Object.fromEntries(ID_REGISTRY.map((id) => [id.prefix, { prefix: id.prefix, label: id.label }]));
+})();
+
+export const BY_LABEL = ((): IdRegistry => {
+  return Object.fromEntries(
+    ID_REGISTRY.map((id) => [id.label.replace(/ /g, '_').toUpperCase(), { prefix: id.prefix, label: id.label }])
+  );
+})();
+
+export function validateId(idObj: Many<IdRegistryValue>, value: string): void {
+  if (!validateIdNoThrow(idObj, value)) {
+    throw messages.createError('invalidIdOrAlias', [
+      Array.isArray(idObj) ? idObj.map((e) => e.label).join(' or ') : idObj.label,
+      value,
+      Array.isArray(idObj) ? idObj.map((e) => e.prefix).join(' or ') : idObj.prefix,
+    ]);
+  }
+}
+export function validateIdNoThrow(idObj: Many<IdRegistryValue>, value) {
+  if (!value || (value.length !== 15 && value.length !== 18)) {
+    return false;
+  }
+  return Array.isArray(idObj) ? idObj.some((e) => value.startsWith(e.prefix)) : value.startsWith(idObj.prefix);
+}
+export function validateVersionNumber(
+  versionNumberString: string,
+  supportedBuildNumberToken: string,
+  supportedBuildNumberToken2: string
+): string {
+  const versionNumber = VersionNumber.from('1.1.0.NEXT');
+  // build number can be a number or valid token
+  if (
+    Number.isNaN(parseInt(versionNumber.build, 10)) &&
+    versionNumber.build !== supportedBuildNumberToken &&
+    versionNumber.build !== supportedBuildNumberToken2
+  ) {
+    if (supportedBuildNumberToken2) {
+      throw messages.createError('errorInvalidBuildNumberForKeywords', [
+        versionNumberString,
+        supportedBuildNumberToken,
+        supportedBuildNumberToken2,
+      ]);
+    } else {
+      throw messages.createError('errorInvalidBuildNumber', [versionNumberString, supportedBuildNumberToken]);
+    }
+  }
+
+  return versionNumberString;
+}
+export async function validatePatchVersion(
+  connection: Connection,
+  versionNumberString: string,
+  packageId: string
+): Promise<void> {
+  const query = `SELECT ContainerOptions FROM Package2 WHERE id ='${packageId}'`;
+  const queryResult = await connection.tooling.query(query);
+
+  if (queryResult.records === null || queryResult.records.length === 0) {
+    throw messages.createError('errorInvalidPackageId', [packageId]);
+  }
+
+  // Enforce a patch version of zero (0) for Locked packages only
+  if (queryResult.records[0].ContainerOptions === 'Locked') {
+    const versionNumber = VersionNumber.from(versionNumberString);
+    if (versionNumber.patch !== '0') {
+      throw messages.createError('errorInvalidPatchNumber', [versionNumberString]);
+    }
+  }
+}
+
+// TODO: let's get rid of this in favor of SfdcUrl.isValidUrl()
+// check that the provided url has a valid format
+export function validUrl(url: string): boolean {
```

Comment:
Covering the preferred function with old function so it makes it simpler to replace when all code has been moved over.